src/gt4py/next/iterator/embedded.py

```diff
@@ -1114,6 +1102,108 @@ def impl(it: ItIterator) -> ItIterator:
     return impl
 
 
+DT = TypeVar("DT")
+
+
+class _List(tuple, Generic[DT]):  # TODO encode size in type?
+    ...
+
+
+@dataclasses.dataclass(frozen=True)
+class _ConstList(Generic[DT]):
+    value: DT
+
+    def __getitem__(self, _):
+        return self.value
+
+
+@builtins.neighbors.register(EMBEDDED)
+def neighbors(offset: runtime.Offset, it: ItIterator) -> _List:
+    # TODO offset provider should be part of the iterator interface?
+    offset_str = offset.value if isinstance(offset, runtime.Offset) else offset
+    return _List(
+        builtins.deref(it.shift(offset_str, i))
+        if builtins.can_deref(it.shift(offset_str, i))
+        else None
+        for i in range(it.offset_provider[offset_str].max_neighbors)  # type: ignore
```

Comment:
Use an assignment expression to avoid repeating expressions:
```suggestion
        builtins.deref(shifted_it)
        for i in range(it.offset_provider[offset_str].max_neighbors)  # type: ignore
        if builtins.can_deref(shifted_it := (it.shift(offset_str, i)))
```